[WIP] Added glob to support wildcard model folder structure

derbaeuerle opened the pull request:
I tried to use hapi-bookshelf-models in my hapi application but with the lack of support of multidimensional folder structure for the models and wildcard model name definition there would was no reason for me to use this library.
With these changes you'll be able to use hapi-bookshelf-models as you know it or with support of an options.models array/string using glob syntax.

If there is anything wrong inside the code or opportunities for improvement, please let me know and i'll fix it.

Commit:
Added glob to support wildcard model folder structure

robinjoseph08 commented on lib/index.js:
```diff
@@ -59,3 +77,31 @@ exports.register.attributes = {
+  }
+};
+
+internals._addModel = function (model) {
```
Since you only use this function once and it's only one line, you can just call `internals.modelObjects.push(model);` directly.

robinjoseph08 commented on lib/index.js:
```diff
@@ -18,38 +24,50 @@ exports.register = function (server, options, next) {
+    }
+  }
+
+  internals._loadModels(internals.bookshelf);
```
`_loadModels` doesn't take an argument

robinjoseph08 commented on lib/index.js:
```diff
@@ -1,13 +1,19 @@
     knex: Joi.object().required(),
     plugins: Joi.array().includes(Joi.string()).default([]),
-    models: Joi.string().required(),
+    models: Joi.alternatives().try(Joi.string(), Joi.array()).required(),
```
I think the better solution here would be to have `Joi.array().single().items(Joi.string()).required()`.

This would make it so the user can pass in either an array or a single item, but Joi would convert it into an array so you don't have to check it's type, you can just iterate over it. This will mean that you'll have to change the `Joi.assert` to a `Joi.validate` on line 22 as well.

This will allow for the case of a user passing in a array of non-glob directories (which you should have a test case for). Right now, your setup only allows for:
1. a single directory
2. a single glob
3. an array of globs

robinjoseph08 commented:
Thanks for the contribution @derbaeuerle! Sorry for such a delayed reply. I'm open to discuss my comments if you don't think they're valid!

leore commented:
@derbaeuerle do you plan on making the changes proposed above?

derbaeuerle commented:
Sorry, I was on vacation for a longer period.
I'll try to adopt the proposed changes in the next 1 or 2 weeks.

derbaeuerle commented:
Hey @robinjoseph08 and @leore,

I implemented the proposed changes and also updated the version of Joi to 7.0.1 to support the new schema description.

leore commented:
@derbaeuerle can you rebase?

derbaeuerle commented:
@leore Of course. The Fork should be up-to-date now. It seems though that the new Joi version is issuing some troubles. I'll check which version should work with the new schema without a failed travis build.

derbaeuerle commented:
The travis build works fine now with an older version of Joi.

robinjoseph08 commented:
@derbaeuerle Sorry, I think there was a bit of a misunderstanding. The reason to use the `.single` function in Joi is to be able to utilize the [`.validate` function](https://github.com/hapijs/joi/blob/master/API.md#validatevalue-schema-options-callback) which coerces the input into a desired output.

So now that you are using `.single`, you can change line 22 to `.validate` instead. This way, you don't have to ever check `typeof options.models` because you will always know that it is an array since Joi will force it into one even if it isn't.

This is to accomplish two outcomes:
1. It simplifies the code since you have less branches/statements.
2. It allows you to cover all possible permutations of `models` being a string/array and glob/non-glob.
   - string of glob
   - string of non-glob
   - array of globs
   - array of non-globs

Your code only supports the first 3 permutations. You could add more code to support the fourth but chances are it'll be repetitive. The better alternative is to use `.validate` to make sure `models` is an array, iterate over that array, and for each element, handle the cases of if it's a glob and if it's not.

Also don't forget to add the test cases for those permutations.

I hope that makes a bit more sense. Feel free to ask if it still doesn't!

derbaeuerle commented:
Hey @robinjoseph08 
Sorry for the misunderstanding. I changed the function and simplified the file analysis.
The option now supports all permutations of the following models:
- string of glob
- string of non-glob
- array of globs
- array of non-globs

robinjoseph08 commented on lib/index.js:
```diff
@@ -1,55 +1,65 @@
+    internals.modelObjects = internals.modelObjects.concat(gFiles);
+  }
+
+  internals._loadModels();
```
Since `_loadModels` is only called once now, I don't think it needs to be extracted out to a function anymore.

robinjoseph08 commented on lib/index.js:
```diff
@@ -1,55 +1,65 @@
+var glob  = require('glob');
 var path  = require('path');
 
+var internals = {};
```
Since the internal functions have been removed, I think the `internals` object isn't necessary anymore.

robinjoseph08 commented on lib/index.js:
```diff
@@ -1,55 +1,65 @@
-    return next(ex);
+  var result = Joi.validate(options, schema);
+  if (result.error) {
+    next(result.error);
```
`return next(result.error);` so that execution stops.

robinjoseph08 commented on test/index.test.js:
```diff
@@ -72,11 +72,215 @@ describe('bookshelf plugin', function () {
+    ], function (err) {
+      expect(err).to.be.undefined;
+      expect(server.plugins.bookshelf.model('User')).to.be.a('function');
+        expect(server.plugins.bookshelf.model('Role')).to.be.a('function');
```
Can you align this line with the rest of the `expect`s?

leore commented:
@robinjoseph08 is this ready for review?

robinjoseph08 commented:
@derbaeuerle Awesome, it all looks good!

If you could squash all of your commits down into one, it'll be ready to be merged!

derbaeuerle commented:
Hey @robinjoseph08,

I squashed all commits into one and pushed the branch so it can be pulled into this project.

robinjoseph08 commented:
@derbaeuerle perfect! Thanks for help and cooperation! I'll be cutting a new version later today :+1:
